core/src/services/memcached/binary.rs

```diff
@@ -90,18 +96,200 @@ impl PacketHeader {
     }
 }
 
+pub enum Connection {
+    Tls(TlsConnection),
+    Tcp(TcpConnection),
+}
+
+impl Connection {
+    pub async fn version(&mut self) -> Result<String> {
+        match self {
+            Self::Tls(conn) => conn.version().await,
+            Self::Tcp(conn) => conn.version().await,
+        }
+    }
+    pub async fn get(&mut self, key: &str) -> Result<Option<Vec<u8>>> {
+        match self {
+            Self::Tls(conn) => conn.get(key).await,
+            Self::Tcp(conn) => conn.get(key).await,
+        }
+    }
+
+    pub async fn set(&mut self, key: &str, val: &[u8], expiration: u32) -> Result<()> {
+        match self {
+            Self::Tls(conn) => conn.set(key, val, expiration).await,
+            Self::Tcp(conn) => conn.set(key, val, expiration).await,
+        }
+    }
+    pub async fn delete(&mut self, key: &str) -> Result<()> {
+        match self {
+            Self::Tls(conn) => conn.delete(key).await,
+            Self::Tcp(conn) => conn.delete(key).await,
+        }
+    }
+}
+
 pub struct Response {
     header: PacketHeader,
     _key: Vec<u8>,
     _extras: Vec<u8>,
     value: Vec<u8>,
 }
 
-pub struct Connection {
+pub struct TlsConnection {
+    io: BufReader<TlsStream<TcpStream>>,
+}
+
+impl TlsConnection {
+    pub fn new(io: TlsStream<TcpStream>) -> Self {
+        Self {
+            io: BufReader::new(io),
+        }
+    }
+
+    pub async fn auth(&mut self, username: &str, password: &str) -> Result<()> {
```

Comment:
It seems like we're repeating a significant amount of code. Perhaps we could work directly with `AsyncWrite`? I believe both `TlsStream<TcpStream>` and `TcpStream` have implemented it.